semantickernel-experimental/src/main/java/com/microsoft/semantickernel/agents/AgentChat.java

```diff
@@ -0,0 +1,205 @@
+                return synchronizeChannelAsync(channelHash)
+                    .flatMap(channel -> channel.getHistoryAsync());    
+            }
+        } finally {
+            lock.unlock();  
+        }
+    }
+
+    /**
+     * Append a message to the conversation. Adding a message while an agent is active is not allowed.
+     *
+     * @param message A non-system message with which to append to the conversation.
+     * @throws KernelException if a system message is present, without taking any other action
+     * @throws KernelException chat has current activity.
+     */
+    public void addChatMessage(ChatMessageContent<?> message) {
+        addChatMessages(Arrays.asList(message));
+    }
+
+    /**
+     * Append messages to the conversation. Adding messages while an agent is active is not allowed.
+     *
+     * @param messages Set of non-system messages with which to append to the conversation.
+     * @throws KernelException if a system message is present, without taking any other action
+     * @throws KernelException chat has current activity.
+     */
+    public void addChatMessages(List<ChatMessageContent<?>> messages) {
+        lock.lock();
+        try {
+           
+            if (messages.stream().anyMatch(it -> it.getAuthorRole() == AuthorRole.SYSTEM)) {
+                throw new SKException(
+                        String.format("History does not support messages with Rople of %s", AuthorRole.SYSTEM));
+            }
+            
+            // Append chat history
+            chatHistory.addAll(messages);
+
+            // Broadcast message to other channels (in parallel)
+            // Note: Able to queue messages without synchronizing channels.
+            List<ChannelReference> channelRefs = 
+                agentChannels.entrySet().stream()
+                .map(entry -> new ChannelReference(entry.getValue(), entry.getKey()))
+                .collect(Collectors.toList());
+
+            broadcastQueue.enqueue(channelRefs, messages);
+
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Construct a new {@link AgentChat} instance.
+     */
+    protected AgentChat() {
+        broadcastQueue = new BroadcastQueue();
+        agentChannels = new HashMap<>();
+        channelMap = new HashMap<>();
+        chatHistory = new ChatHistory();
+        lock = new ReentrantLock();
+    }
+
+    /**
+     * Exposes the internal history to subclasses.
+     */
+    protected ChatHistory getHistory() {
+        return chatHistory;
+    }
+
+    /**
+     * Process a discrete incremental interaction between a single Agent an a AgentChat.
+     *
+     * @param agent The agent actively interacting with the chat.
+     * @return Asynchronous enumeration of messages.
+     */
+    protected Mono<List<ChatMessageContent<?>>> invokeAgentAsync(Agent agent) {
+        lock.lock();
+        try {
+            return getOrCreateChannel(agent)
```

Comment:
@johnoliver - need your input on this port from https://github.com/microsoft/semantic-kernel/blob/7e2bca7c98ec7b38babc266267d5f0d1960a508b/dotnet/src/Agents/Abstractions/AgentChat.cs#L196 
I'm not confident about the reactor stream.